src/chainstate/stacks/db/blocks.rs

```diff
@@ -5061,15 +5063,73 @@ impl StacksChainState {
         }
     }
 
+    /// Check if current PoX reward cycle (as of `burn_tip_height`) has handled any
+    ///  Clarity VM work necessary at the start of the cycle (i.e., processing of accelerated unlocks
+    ///  for failed stackers).
+    /// If it has not yet been handled, then perform that work now.
+    fn check_and_handle_reward_start(
+        burn_tip_height: u64,
+        burn_dbconn: &dyn BurnStateDB,
+        sortition_dbconn: &dyn SortitionDBRef,
+        clarity_tx: &mut ClarityTx,
+        chain_tip: &StacksHeaderInfo,
+        parent_sortition_id: &SortitionId,
+    ) -> Result<Vec<StacksTransactionEvent>, Error> {
+        let mut pox_reward_cycle = Burnchain::static_block_height_to_reward_cycle(
+            burn_tip_height,
+            burn_dbconn.get_burn_start_height().into(),
+            burn_dbconn.get_pox_reward_cycle_length().into(),
+        ).expect("FATAL: Unrecoverable chainstate corruption: Epoch 2.1 code evaluated before first burn block height");
+        // Do not try to handle auto-unlocks on pox_reward_cycle 0
+        // This cannot even occur in the mainchain, because 2.1 starts much
+        //  after the 1st reward cycle, however, this could come up in mocknets or regtest.
+        if pox_reward_cycle > 1 {
+            if Burnchain::is_before_reward_cycle(
```

Comment:
Given the way this method is used, I just want to understand a few things:

* `burn_tip_height` always increases by 1 between subsequent calls to `check_and_handle_reward_start()`, right?  There's no way that we could process a reward cycle and somehow _skip_ this `if` statement's "true" block?  Asking because the condition `(effective_height % reward_cycle_length) <= 1` suggests that there this block will be executed more than once per reward cycle -- i.e. when the modulus is 0 and when it is 1?

* If this method is supposed to be true only when `burn_tip_height` is the burn block height refers to a block that was mined just before a new reward cycle begins -- i.e. it's the last block in the previous reward cycle, might the method name suggest this?  For example, `is_last_reward_cycle_block()` might indicate this better.  Up to you.

Reply:
> burn_tip_height always increases by 1 between subsequent calls to check_and_handle_reward_start(), right? There's no way that we could process a reward cycle and somehow skip this if statement's "true" block? Asking because the condition (effective_height % reward_cycle_length) <= 1 suggests that there this block will be executed more than once per reward cycle -- i.e. when the modulus is 0 and when it is 1?

It depends on what your definition of time is for "subsequent calls". This call is checking whether or not the current reward cycle has been "handled" in this Stacks fork. The "current" reward cycle is either the active reward cycle if the reward set has already been stored (`> mod 1`) or the previous reward cycle if the active reward cycle hasn't stored its reward set yet. If Stacks blocks proceed without forks or missed burn blocks, then yes, the burn height increments by 1 on every invocation, and the reward cycle will be "handled" just once. However, if there's multiple Stacks forks, it may be the case that a fork needs to handle the reward cycle again.

> If this method is supposed to be true only when burn_tip_height is the burn block height refers to a block that was mined just before a new reward cycle begins -- i.e. it's the last block in the previous reward cycle, might the method name suggest this? For example, is_last_reward_cycle_block() might indicate this better. Up to you.

Because this method is checking `<= mod 1`, it returns true for two different block heights in each reward cycle. It's trying to figure out whether or not the reward set has already been written by the sortition and chainstate databases -- that occurs at modulus block 1.

Reply:
> Because this method is checking <= mod 1, it returns true for two different block heights in each reward cycle. It's trying to figure out whether or not the reward set has already been written by the sortition and chainstate databases -- that occurs at modulus block 1.

This is where I think I'm confused.  Why two, and why these two?

Suppose I have a reward cycle in which PoX is active, but where the first Stacks block in this cycle gets mined in the middle of the cycle (so, `burn_tip_height % reward_cycle_len == 1050`).  Then, once this code is reached, the `if-true` block will not get processed, meaning that `pox_reward_cycle -= 1` will _not_ get processed, which in turn means that anyone who clinched zero reward slots for this reward cycle will _not_ unlock in this block (since `pox_reward_cycle` refers to the now-current reward cycle, instead of the previous reward cycle).  Correct me if I'm wrong, but won't this mean that users who earned zero reward slots in the previous reward cycle will remain locked in the current cycle?

More generally, what would happen if users locked their tokens for cycle `N - 1`, but a fork was constructed in cycle `N + k` for some `k >= 1` that skipped all of the blocks in cycle `N`...`N + k - 1` but mined off of a block that descended from cycle `N`'s anchor block (which would have been mined towards the end of cycle `N - 1`)?  It looks like these users would remain locked in this fork, because the code would never call `Self::handle_pox_cycle_start()` for any cycle `N` ... `N + k - 1`?  If so, then I don't think this is what we want?

Reply:
The answer right now is that `B3` won't process those unlocks. The current implementation will only try to process unlocks for the most recently processed reward cycle with an active anchor block in the current Stacks fork.

An alternative would be to fetch the most recently "handled" reward cycle in the current Stacks fork, and then "handle" each reward cycle since that point. This, of course, is unbounded work, but it would be a low bound: really deep forks would need to do lots of checks, but the checks would be short, because no processing occurs for blocks that don't descend from the anchor block (which deep forks necessarily wouldn't).

> This is where I think I'm confused. Why two, and why these two?

The reward set and unlock state for reward cycle `n` is computed at `1 mod R`, and is only written to the sortition database once that sortition has been processed. This branch is telling the handler what the latest reward set it could successfully fetch information for is. The reason that it cannot fetch information for `1 mod R` is that miners need to be able to figure that out without already having processed the sortition. This is _possible_ but it would require that the miners plumb the `pox_start_cycle_info` into the block processing code (this is similar to how they compute the first two addresses to send their commitments to in the reward cycle, but it occurs in different areas of the code).

Reply:
Thanks, I understand now :+1: